File: packages/ts/src/domain/audit.ts
import type { RunId } from "./identifiers";

export const AuditEventType = {
  ModelCalled: "model_called",
  ToolExecuted: "tool_executed",
  ApprovalRequested: "approval_requested",
  ApprovalDecided: "approval_decided",
  RunCompleted: "run_completed",
  RunFailed: "run_failed",
} as const;
export type AuditEventType = (typeof AuditEventType)[keyof typeof AuditEventType];

/** An immutable record of one thing that happened during a run. */
export interface AuditEvent {
  readonly runId: RunId;
  readonly type: AuditEventType;
  /** UTC ISO-8601 timestamp, taken from the Clock port. */
  readonly at: string;
  readonly data: Readonly<Record<string, unknown>>;
}

File: packages/py/src/airlock/domain/audit.py
"""Audit events: an immutable record of everything a run does."""

from enum import StrEnum

from pydantic import BaseModel, ConfigDict

from airlock.domain.identifiers import RunId


class AuditEventType(StrEnum):
    MODEL_CALLED = "model_called"
    TOOL_EXECUTED = "tool_executed"
    APPROVAL_REQUESTED = "approval_requested"
    APPROVAL_DECIDED = "approval_decided"
    RUN_COMPLETED = "run_completed"
    RUN_FAILED = "run_failed"


class AuditEvent(BaseModel):
    model_config = ConfigDict(frozen=True)

    run_id: RunId
    type: AuditEventType
    at: str
    """UTC ISO-8601 timestamp, taken from the Clock port."""
    data: dict[str, object]
